// takumi-js/dist/render-EhI8Ytcl.mjs
import { extractEmojis } from "@takumi-rs/helpers/emoji";
import { extractResourceUrls, fetchResources } from "@takumi-rs/helpers";
import { fromJsx } from "@takumi-rs/helpers/jsx";
import { fromHtml } from "@takumi-rs/helpers/html";
//#region src/renderer.ts
async function loadRendererResources(renderer, options) {
	const tasks = [];
	if (options?.fonts && options.fonts.length > 0) tasks.push(renderer.loadFonts(options.fonts));
	if (options?.persistentImages && options.persistentImages.length > 0) tasks.push(...options.persistentImages.map((image) => Promise.resolve(renderer.putPersistentImage(image, options.signal))));
	if (tasks.length > 0) await Promise.all(tasks);
}
//#endregion
//#region src/import.ts
let importPromise = null;
function getImports(module) {
	importPromise ??= getImportsImpl(module);
	return importPromise;
}
async function getImportsImpl(module) {
	if (module) return initializeWasm(module);
	if (shouldSkipCoreImport()) return initializeWasm(importWasmBindings());
	try {
		return await import("@takumi-rs/core");
	} catch (error) {
		if (isNodeEnvironment()) throw new Error("Failed to load @takumi-rs/core in Node.js runtime. Takumi requires the native napi-rs module in Node environments.", { cause: error });
		console.warn("Unable to import @takumi-rs/core. Falling back to auto-detection of WASM bindings.", { cause: error });
	}
	return initializeWasm(importWasmBindings());
}
async function initializeWasm(module) {
	const wasm = await import("@takumi-rs/wasm");
	const resolvedModule = typeof module === "function" ? await module() : await module;
	const wasmModule = resolvedModule !== void 0 && typeof resolvedModule === "object" && "default" in resolvedModule ? resolvedModule.default : resolvedModule;
	try {
		await wasm.default(wasmModule ? { module_or_path: wasmModule } : void 0);
		return wasm;
	} catch (error) {
		throw new Error("Couldn't automatically resolve Takumi native bindings. Please specify the module option with the WASM module.", { cause: error });
	}
}
function hackFakeProcessForBrowser() {
	const before = globalThis.process;
	globalThis.process ??= {};
	globalThis.process.env ??= {};
	return before;
}
async function importWasmBindings() {
	const beforeProcess = hackFakeProcessForBrowser();
	const nextPath = "@takumi-rs/wasm/next";
	if (process.env.NEXT_RUNTIME) {
		globalThis.process = beforeProcess;
		return import(
			/* @vite-ignore */
			nextPath
);
	}
	globalThis.process = beforeProcess;
	return import(
		/* webpackIgnore: true */
		"@takumi-rs/wasm/auto"
);
}
function shouldSkipCoreImport() {
	const beforeProcess = hackFakeProcessForBrowser();
	if (process.env.NEXT_RUNTIME === "edge") {
		globalThis.process = beforeProcess;
		return true;
	}
	globalThis.process = beforeProcess;
	if (typeof window !== "undefined") return true;
	if (typeof navigator !== "undefined" && navigator.userAgent === "Cloudflare-Workers") return true;
	if ("WebSocketPair" in globalThis) return true;
	if ("EdgeRuntime" in globalThis) return true;
	const maybeWorkerGlobalScope = globalThis.WorkerGlobalScope;
	if (maybeWorkerGlobalScope !== void 0 && maybeWorkerGlobalScope.prototype.isPrototypeOf(globalThis)) return true;
	return false;
}
function isNodeEnvironment() {
	return typeof process !== "undefined" && typeof process.versions === "object" && process.versions !== null && typeof process.versions.node === "string";
}
//#endregion
//#region src/render.ts
let globalRenderer;
function transformElement(element, options) {
	if (typeof element === "string") return fromHtml(element);
	return fromJsx(element, options?.jsx);
}
async function render(element, options) {
	const imports = await getImports(options && "module" in options ? options.module : void 0);
	const isExternalRenderer = options && "renderer" in options;
	const renderer = isExternalRenderer ? options.renderer : globalRenderer ??= new imports.Renderer({ loadDefaultFonts: options?.loadDefaultFonts });
	if (!isExternalRenderer) await loadRendererResources(renderer, options);
	const { node: originalNode, stylesheets } = await transformElement(element, options);
	const emojiType = options?.emoji ?? "twemoji";
	const node = emojiType !== "from-font" ? extractEmojis(originalNode, emojiType) : originalNode;
	const fetchedResources = options?.fetchedResources ?? await fetchResources(extractResourceUrls(node), options?.resourcesOptions);
	const renderOptions = {
		...options,
		fetchedResources,
		stylesheets: [...options?.stylesheets ?? [], ...stylesheets]
	};
	return renderer.render(node, renderOptions, options?.signal);
}
//#endregion
export { render as t };
