// takumi-js/dist/response.mjs
import { t as render } from "./render-EhI8Ytcl.mjs";
//#region src/response/index.ts
function mergeOptions(defaultOptions, options) {
	if (!defaultOptions) return options;
	if (!options) return defaultOptions;
	const headers = new Headers(defaultOptions?.headers);
	if (options?.headers) new Headers(options.headers).forEach((value, key) => {
		headers.set(key, value);
	});
	return {
		...defaultOptions,
		...options,
		headers,
		stylesheets: [...defaultOptions?.stylesheets ?? [], ...options?.stylesheets ?? []]
	};
}
const contentTypeMap = {
	png: "image/png",
	jpeg: "image/jpeg",
	webp: "image/webp",
	ico: "image/x-icon",
	raw: "application/octet-stream"
};
function defaultErrorHandler(error) {
	console.error("Failed to render image.");
	console.error(error);
}
function createImageResponse(defaultOptions) {
	return function imageResponse(element, options) {
		const mergedOptions = mergeOptions(defaultOptions, options);
		const { promise: ready, reject: rejectReady, resolve: resolveReady } = Promise.withResolvers();
		const stream = new ReadableStream({
			type: "bytes",
			async start(controller) {
				try {
					const image = await render(element, mergedOptions);
					controller.enqueue(image);
					controller.close();
					resolveReady();
				} catch (error) {
					controller.error(error);
					rejectReady(error);
					await (mergedOptions?.onError ?? defaultErrorHandler)(error);
				}
			}
		});
		const headers = new Headers(mergedOptions?.headers);
		if (!headers.get("content-type")) headers.set("content-type", contentTypeMap[mergedOptions?.format ?? "png"]);
		const response = new Response(stream, {
			headers,
			status: mergedOptions?.status,
			statusText: mergedOptions?.statusText
		});
		return Object.defineProperty(response, "ready", {
			enumerable: false,
			value: ready,
			writable: false
		});
	};
}
let defaultImageResponse;
var ImageResponse = class extends Response {
	ready;
	constructor(component, options) {
		defaultImageResponse ??= createImageResponse();
		const response = defaultImageResponse(component, options);
		super(response.body, response);
		this.ready = response.ready;
	}
};
//#endregion
export { ImageResponse, ImageResponse as default, createImageResponse };
