// fumadocs-core/dist/toc.js
"use client";
import { createContext, use, useEffect, useMemo, useRef, useState } from "react";
import { jsx } from "react/jsx-runtime";
import scrollIntoView from "scroll-into-view-if-needed";
//#region src/utils/merge-refs.ts
function mergeRefs(...refs) {
	return (value) => {
		refs.forEach((ref) => {
			if (typeof ref === "function") ref(value);
			else if (ref != null) ref.current = value;
		});
	};
}
//#endregion
//#region src/toc.tsx
const ItemsContext = createContext(null);
const ScrollContext = createContext(null);
/** Optional: add auto-scroll to TOC items. */
function ScrollProvider({ containerRef, children }) {
	return /* @__PURE__ */ jsx(ScrollContext, {
		value: containerRef,
		children
	});
}
function AnchorProvider({ toc, single = false, children }) {
	const observer = useMemo(() => new Observer(), []);
	const [items, setItems] = useState(observer.items);
	observer.single = single;
	useEffect(() => {
		observer.setItems(toc);
	}, [observer, toc]);
	useEffect(() => {
		observer.watch({
			rootMargin: "0px",
			threshold: .98
		});
		observer.onChange = () => setItems(observer.items);
		return () => {
			observer.unwatch();
		};
	}, [observer]);
	return /* @__PURE__ */ jsx(ItemsContext, {
		value: items,
		children
	});
}
function TOCItem({ ref, onActiveChange = () => null, ...props }) {
	const id = props.href ? getItemId(props.href) : null;
	const items = useItems();
	const itemData = id ? items.find((item) => item.id === id) : null;
	const containerRef = use(ScrollContext);
	const isInitialMountRef = useRef(true);
	const anchorRef = useRef(null);
	const [active, setActive] = useState(() => itemData != null && itemData.active);
	if (itemData && itemData.active !== active) {
		setActive(itemData.active);
		onActiveChange(itemData.active);
	}
	const shouldScroll = itemData?.active && items.every((item) => !item.active || item.t <= itemData.t);
	useEffect(() => {
		const anchor = anchorRef.current;
		const container = containerRef?.current;
		if (shouldScroll && container && anchor) scrollIntoView(anchor, {
			behavior: isInitialMountRef.current ? "instant" : "smooth",
			block: "center",
			inline: "center",
			scrollMode: "always",
			boundary: container
		});
		isInitialMountRef.current = false;
	}, [shouldScroll, containerRef]);
	return /* @__PURE__ */ jsx("a", {
		ref: mergeRefs(anchorRef, ref),
		"data-active": active,
		...props
	});
}
/**
* The estimated active heading ID
*/
function useActiveAnchor() {
	const items = useItems();
	return useMemo(() => {
		let out;
		for (const item of items) {
			if (!item.active) continue;
			if (!out || item.t > out.t) out = item;
		}
		return out?.id;
	}, [items]);
}
/**
* The id of visible anchors
*/
function useActiveAnchors() {
	const items = useItems();
	return useMemo(() => {
		const out = [];
		for (const item of items) if (item.active) out.push(item.id);
		return out;
	}, [items]);
}
function useItems() {
	const ctx = use(ItemsContext);
	if (!ctx) throw new Error(`Component must be used under the <AnchorProvider /> component.`);
	return ctx;
}
function getItemId(url) {
	if (url.startsWith("#")) return url.slice(1);
	return null;
}
var Observer = class {
	constructor() {
		this.items = [];
		this.single = false;
		this.observer = null;
	}
	callback(entries) {
		if (entries.length === 0) return;
		let hasActive = false;
		this.items = this.items.map((item) => {
			const entry = entries.find((entry) => entry.target.id === item.id);
			let active = entry ? entry.isIntersecting : item.active && !item.fallback;
			if (this.single && hasActive) active = false;
			if (item.active !== active) item = {
				...item,
				t: Date.now(),
				active,
				fallback: false
			};
			if (active) hasActive = true;
			return item;
		});
		if (!hasActive && entries[0].rootBounds) {
			const viewTop = entries[0].rootBounds.top;
			let min = Number.MAX_VALUE;
			let fallbackIdx = -1;
			for (let i = 0; i < this.items.length; i++) {
				const element = document.getElementById(this.items[i].id);
				if (!element) continue;
				const d = Math.abs(viewTop - element.getBoundingClientRect().top);
				if (d < min) {
					fallbackIdx = i;
					min = d;
				}
			}
			if (fallbackIdx !== -1) this.items[fallbackIdx] = {
				...this.items[fallbackIdx],
				active: true,
				fallback: true,
				t: Date.now()
			};
		}
		this.onChange?.();
	}
	setItems(newItems) {
		const observer = this.observer;
		if (observer) for (const item of this.items) {
			const element = document.getElementById(item.id);
			if (!element) continue;
			observer.unobserve(element);
		}
		this.items = [];
		for (const item of newItems) {
			const id = getItemId(item.url);
			if (!id) continue;
			this.items.push({
				id,
				active: false,
				fallback: false,
				t: 0,
				original: item
			});
		}
		this.watchItems();
	}
	watch(options) {
		if (this.observer) return;
		this.observer = new IntersectionObserver(this.callback.bind(this), options);
		this.watchItems();
	}
	watchItems() {
		if (!this.observer) return;
		for (const item of this.items) {
			const element = document.getElementById(item.id);
			if (!element) continue;
			this.observer.observe(element);
		}
	}
	unwatch() {
		this.observer?.disconnect();
		this.observer = null;
	}
};
//#endregion
export { AnchorProvider, ScrollProvider, TOCItem, useActiveAnchor, useActiveAnchors, useItems };
