// @tanstack/devtools/dist/index.js
export { PLUGIN_CONTAINER_ID, PLUGIN_TITLE_CONTAINER_ID } from './chunk/A767CXXU.js';
import { initialState } from './chunk/BNGI36V3.js';

// src/core.ts
var TanStackDevtoolsCore = class {
  #config = {
    ...initialState.settings
  };
  #plugins = [];
  #isMounted = false;
  #isMounting = false;
  #abortMount = false;
  #dispose;
  #eventBus;
  #eventBusConfig;
  #setPlugins;
  constructor(init) {
    this.#plugins = init.plugins || [];
    this.#eventBusConfig = init.eventBusConfig;
    this.#config = {
      ...this.#config,
      ...init.config
    };
  }
  mount(el) {
    if (typeof document === "undefined") return;
    if (this.#isMounted || this.#isMounting) {
      throw new Error("Devtools is already mounted");
    }
    this.#isMounting = true;
    this.#abortMount = false;
    import('./mount-impl/Z6LKUI5N.js').then(({ mountDevtools }) => {
      if (this.#abortMount) {
        this.#isMounting = false;
        return;
      }
      const result = mountDevtools({
        el,
        plugins: this.#plugins,
        config: this.#config,
        eventBusConfig: this.#eventBusConfig,
        onSetPlugins: (setPlugins) => {
          this.#setPlugins = setPlugins;
        }
      });
      this.#dispose = result.dispose;
      this.#eventBus = result.eventBus;
      this.#isMounted = true;
      this.#isMounting = false;
    }).catch((err) => {
      this.#isMounting = false;
    });
  }
  unmount() {
    if (!this.#isMounted && !this.#isMounting) {
      throw new Error("Devtools is not mounted");
    }
    if (this.#isMounting) {
      this.#abortMount = true;
      this.#isMounting = false;
      return;
    }
    this.#eventBus?.stop();
    this.#dispose?.();
    this.#isMounted = false;
  }
  setConfig(config) {
    this.#config = {
      ...this.#config,
      ...config
    };
    if (config.plugins) {
      this.#plugins = config.plugins;
      if (this.#isMounted && this.#setPlugins) {
        this.#setPlugins(config.plugins);
      }
    }
  }
};

export { TanStackDevtoolsCore };
